### DISCUSSION TITLE
blocking i2s function without DMA

### DISCUSSION THREAD

**Discussion by proleaveprod** [pschatzmann/arduino-audio-tools#754] [Q&A] [answered]
Hi! 
Is there way to write data to i2s without DMA? 

I need to generate signal impulse in **real-time**, currently, i need to generate my array at once, when i call the generate function

kit.write() is not matched, cause it use DMA and it uses ring-buffer of DMA. 

If i use kit.write() , my signal generates, but delay between calling of kit.write() and real generating on output of es8388.

I tried to start and stop i2s by function i2s_start and i2s_stop and it worked in realtime, but 1 in 16 impulses i hear strange clip and i cant understand why it comes.  

Please tell me...

**Comment by pschatzmann** [comment, marked as answer]
You can minimize the delay by decreasing the i2s buffer. 
You might get rid of the strange noise by setting the use_apll and tx_desc_auto_clear to true
